File: Week03/LeetCode-0001.cpp
class Solution {
public:
    /// O(NlogN) algorithm with O(N) extra space
    vector<int> twoSum(vector<int>& nums, int target) {
        // O(N) to create the temporary array
        vector<pair<int, int>> A;
        for (int i=0; i<nums.size(); i++)
            A.push_back({nums[i], i});
        
        // NlogN to sort
        std::sort(A.begin(), A.end());
        
        // O(N) to find the pair
        int i=0;
        int j=A.size()-1;
        while (i < j){
            int total = A[i].first+A[j].first;
            if (total == target) return {A[i].second, A[j].second};
            else if (total < target) i++;
            else j--;
        } // while
        
        return {-1, -1};
    } // end-twoSum
};

File: Week03/LeetCode-0001.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """
        O(NlogN) algorithm with O(N) extra space
        """
        # O(N) to create the temporary array
        A = []
        for i in range(len(nums)):
            A.append([nums[i], i])
        
        # NlogN to sort
        A.sort()
        
        # O(N) to find the pair
        i = 0
        j = len(A)-1
        while i < j:
            total = A[i][0] + A[j][0]
            if total == target: 
                return [A[i][1], A[j][1]]
            elif total < target:
                i += 1
            else:
                j -= 1        
        return [-1, -1]  # No such pair
        